src/console/Program.cs

```diff
@@ -27,47 +31,91 @@ public class Program
         public static async Task Main()
         {
             var config = GetEnvironmentConfig();
+            var eventHubName = GetEventHubName(config);
 
-            // determine which event hub to read from
-            var eventHub = Environment.GetEnvironmentVariable("WEBJOBS_NAME");
-            if (eventHub == null)
-            {
-                eventHub = config.GetSection("Console:EventHub").Value;
-            }
-
-            System.Console.WriteLine($"Reading from event hub: {eventHub}");
-            System.Console.WriteLine($"Logs and Metrics will be written to Application Insights");
-            var eventHubOptions = GetEventHubInfo(config, eventHub);
+            var eventHubOptions = GetEventHubInfo(config, eventHubName);
 
             EnsureArg.IsNotNullOrWhiteSpace(eventHubOptions.EventHubConnectionString);
             EnsureArg.IsNotNullOrWhiteSpace(eventHubOptions.EventHubName);
 
-            var eventBatchingOptions = new EventBatchingOptions();
-            config.GetSection(EventBatchingOptions.Settings).Bind(eventBatchingOptions);
+            var serviceCollection = GetRequiredServiceCollection(config, eventHubName);
+            var serviceProvider = serviceCollection.BuildServiceProvider();
+
+            var credentialOptions = new CredentialOptions();
+            credentialOptions.SystemAssignedManagedIdentity = true;
+            var azureCredentialProvider = new AzureCredentialProvider(credentialOptions);
+            var managedIdentityCredentialService = new AzureCredentialService(azureCredentialProvider);
 
-            var serviceProvider = GetRequiredServiceProvider(config, eventHub);
             var logger = serviceProvider.GetRequiredService<ITelemetryLogger>();
-            var eventConsumers = GetEventConsumers(config, eventHub, serviceProvider, logger);
+
+            var eventConsumers = GetEventConsumers(config, managedIdentityCredentialService, eventHubName, serviceProvider, logger);
 
             var storageOptions = new StorageCheckpointOptions();
             config.GetSection(StorageCheckpointOptions.Settings).Bind(storageOptions);
-            storageOptions.BlobPrefix = eventHub;
```

Comment:
I wonder is there is a pattern here that would provide more flexibility in the future. What about creating an interface IStorageCheckpointClientProvider that has a GetClient(IConfiguration config). This would allow for different implementations of IStorageCheckpointClientProvider, the default could be for MI, but a consumer could provide an implementation that still uses a connection string.

This will provide flexibility by allowing consumers to create their own appsettings.json scheme that works with their own implementation of IStorageCheckpointClientProvider

This pattern could be used for other clients.

Reply:
Yeah, that is a good idea. I ended up moving things around which should allow the customer to have more flexibility. I decided to create a BlobContainerClientFactory which can take an overload of IAzureCredentialProvider which can construct to blob client with some custom credentials implementation. 

BlobContainerClientFactory will also by default support default azure credential and connection string, and the user can toggle between those by changing the config environment variables.